File: 250508_Opt/Matrix_IJ/matrix_ij.f90
program memory_access
  implicit none
  integer, parameter :: N = 4096
  real(8), dimension(N, N) :: A
  real(8) :: sum
  integer :: i, j
  real(8) :: t_start, t_end

  ! Initialize matrix
  do i = 1, N
    do j = 1, N
      A(i, j) = (i + j) * 0.5d0
    end do
  end do

  ! Column-major access (cache-friendly in Fortran)
  call cpu_time(t_start)
  sum = 0.0d0
  do j = 1, N
    do i = 1, N
      sum = sum + A(i, j)
    end do
  end do
  call cpu_time(t_end)
  print *, "Column-major sum:", sum
  print *, "Column-major time:", t_end - t_start, "sec"

  ! Row-major access (cache-unfriendly in Fortran)
  call cpu_time(t_start)
  sum = 0.0d0
  do i = 1, N
    do j = 1, N
      sum = sum + A(i, j)
    end do
  end do
  call cpu_time(t_end)
  print *, "Row-major sum:", sum
  print *, "Row-major time:", t_end - t_start, "sec"

end program memory_access

File: 250508_Opt/Matrix_IJ/matrix_ij.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define N 4096

double A[N][N];

// Row-major traversal (cache-friendly)
void row_major_access() {
    double sum = 0.0;
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            sum += A[i][j];
    printf("Row-major sum: %f\n", sum);
}

// Column-major traversal (cache-unfriendly in C)
void col_major_access() {
    double sum = 0.0;
    for (int j = 0; j < N; ++j)
        for (int i = 0; i < N; ++i)
            sum += A[i][j];
    printf("Column-major sum: %f\n", sum);
}

int main() {
    // Initialize matrix
    for (int i = 0; i < N; ++i)
        for (int j = 0; j < N; ++j)
            A[i][j] = (i + j) * 0.5;

    clock_t start, end;

    // Measure row-major access time
    start = clock();
    row_major_access();
    end = clock();
    printf("Row-major time: %.3f sec\n", (double)(end - start) / CLOCKS_PER_SEC);

    // Measure column-major access time
    start = clock();
    col_major_access();
    end = clock();
    printf("Column-major time: %.3f sec\n", (double)(end - start) / CLOCKS_PER_SEC);

    return 0;
}

